torchaudio/csrc/sox/effects.cpp

```diff
@@ -50,10 +50,11 @@ void shutdown_sox_effects() {
   }
 }
 
-c10::intrusive_ptr<TensorSignal> apply_effects_tensor(
-    const c10::intrusive_ptr<TensorSignal>& input_signal,
-    std::vector<std::vector<std::string>> effects) {
-  auto in_tensor = input_signal->getTensor();
+std::tuple<torch::Tensor, int64_t> apply_effects_tensor(
+    std::tuple<torch::Tensor, int64_t> input_signal,
+    std::vector<std::vector<std::string>> effects,
+    c10::optional<bool>& channels_first) {
```

Comment:
I think `bool` type instead of `c10::optional<bool>` type better fits the context. All the Python code that calls this function has signature of `channels_first: bool = true`. So it should always have a boolean value.

Reply:
Also could you untangle `std::tuple<torch::Tensor, int64_t> input_signal` into `torch::Tensor waveform` and `int64_t sample_rate`? That way the interface between Python and C++ becomes flat (not need to construct a tuple `(waveform, sample_rate)`) so it looks better.

Then we have option to whether construct a `TensorSignal` object here or apply the same change to `addInputTensor` (and to utility functions used in there like `get_signalinfo`). I will let you decide as the later approach feels digressing from the main point of the PR.